build-packages/ts-server-plugin/src/cache.ts

```diff
@@ -0,0 +1,396 @@
+/* eslint-disable consistent-return */
+import ts from 'typescript/lib/tsserverlibrary';
+
+import { NamespaceDeclaration } from './declaration';
+import { NamespaceReference } from './reference';
+import {
+    CLASS_PLUGIN_METHOD_TYPE, CLASS_PLUGIN_PROPERTY_TYPE, CLASS_PLUGIN_STATIC_TYPE, ClassPluginTypes, FUNCTION_PLUGIN_TYPE
+} from './util/config';
+import { Ctx } from './util/context';
+import { getAllThemeFiles } from './util/parent-theme';
+
+type DeclarationCacheMap = Record<string, NamespaceDeclaration>;
+type ReferenceCacheMap = Record<string, Array<NamespaceReference>>;
+type FileToNamespaceMap = Record<string, Array<NamespaceReference | NamespaceDeclaration>>;
+
+export class Cache {
+    ctx: Ctx;
+
+    mosaicSourceFiles: ts.SourceFile[] | undefined;
+
+    mosaicProgram: ts.Program | undefined;
+
+    hasCachedAll = false;
+
+    declarationMap: DeclarationCacheMap = {};
+
+    referenceMap: ReferenceCacheMap = {};
+
+    fileToNamespaceMap: FileToNamespaceMap = {};
+
+    // (cache) Lookup plugins by namespace
+    // Provide error diagnostics (lookup plugins by namespace)
+    // Update cache by filename
+
+    constructor(
+        ctx: Ctx
+    ) {
+        this.ctx = ctx;
+    }
+
+    getProgramSourceFiles(): readonly ts.SourceFile[] {
+        const program = this.ctx.info.project.getLanguageService().getProgram();
+        if (!program) {
+            return [];
+        }
+        const sourceFiles = program.getSourceFiles();
+        if (!sourceFiles) {
+            return [];
+        }
+
+        return sourceFiles;
+    }
+
+    getSourceFileByPath(fileName: string, tryCount = 0): ts.SourceFile | undefined {
+        // TODO: optimize this function
+
+        const program = this.ctx.info.project.getLanguageService().getProgram();
+        if (!program || tryCount > 1) {
+            return;
+        }
+        // ^^^ Allow only one try to get the source file
+
+        const sourceFile = program.getSourceFile(fileName);
+
+        if (sourceFile) {
+            // this.ctx.info.project.projectService.logger.info(
```

Comment:
Do we want to keep commented lines for logging? or maybe it will be better to keep them? Does this logger log data in the TS Server output logs?